### ISSUE TITLE
clang-tidy false positive cppcoreguidelines assignment operator with explicit object parameters

### ISSUE THREAD

**Issue by cgnitash** [llvm/llvm-project#110798]
When using an assignment operator with explicit object parameters,
 1. clang-tidy mistakes assignment operators as a copy-assignment operator. 
 2. The guideline for returning `*this` is too strict (returning self has the same semantics afaik). 
```
struct S
{
  S &operator=(this S& self, int)
  {
    return self;
  }
};
```
clang-tidy gives
> Warning: operator=() should take 'S const&', 'S&&' or 'S' [cppcoreguidelines-c-copy-assignment-signature]
>    3 |   S &operator=(this S& self, int)
>       |   ^

and

> warning: operator=() should always return '*this' [cppcoreguidelines-c-copy-assignment-signature]
>   5 |     return self;
>      |     ^
